crates/flipperzero/src/furi/message_queue.rs

```diff
@@ -0,0 +1,82 @@
+use core::ffi::c_void;
+use flipperzero_sys::furi::message_queue;
+use crate::furi::Result;
+
+/// MessageQueue provides a safe wrapper around the furi message queue primitive.
+pub struct MessageQueue<M: Sized> {
+    hnd: *const message_queue::FuriMessageQueue,
+    _marker: core::marker::PhantomData<M>,
+}
+
+pub const TIMEOUT_NEVER: u32 = 0xFFFFFFFF; // corresponds to FuriWaitForever
+
+impl<M: Sized> MessageQueue<M> {
+    /// Constructs a message queue with the given capacity.
+    pub fn new(capacity: usize) -> Self {
+        Self {
+            hnd: unsafe { message_queue::alloc(capacity, core::mem::size_of::<M>()) },
+            _marker: core::marker::PhantomData::<M>,
+        }
+    }
+
+    // Attempts to add the message to the end of the queue, waiting up to timeout ticks.
+    pub fn put(&self, msg: M, timeout_ticks: u32) -> Result<()> {
```

Comment:
I just added [`duration_to_ticks`](https://github.com/dcoles/flipperzero-rs/blob/aa7690b733b02a47ee928ed7b74c46db75a045bc/crates/sys/src/furi/kernel.rs#L7) helper that calls `furi_ms_to_ticks` internally (on the current Flipper Zero [this is just a passthrough](https://sourcegraph.com/github.com/flipperdevices/flipperzero-firmware@0f9ea92/-/blob/furi/core/kernel.c?L151) 😁).